Add smithy-based interface for numbergen (previously extras)

stevelr opened the pull request:
add first-draft builtin.smithy for builtin capability providers (wasmcloud:builtin:*)
- logging
- random
- guid
- sequence

Please review apis

autodidaddict commented on builtin/builtin.smithy:
```diff
@@ -0,0 +1,119 @@
+/// These guids are known as "version 4", meaning all bits are random or pseudo-random.
+///
+@wasmbus(
+    contractId: "wasmcloud:builtin:guid",
```
guid, random, and sequence are all part of `wasmcloud:builtin:extras`, what you've got here is a contract-per-operation, which is not what we have today.

autodidaddict commented on builtin/builtin.smithy:
```diff
@@ -0,0 +1,119 @@
+@wasmbus(
+    contractId: "wasmcloud:builtin:random",
+    providerReceive: true )
+service Random {
```
Random should not be a service, it's an operation on the service that implements `wasmcloud:builtin:extras`

autodidaddict commented on builtin/builtin.smithy:
```diff
@@ -0,0 +1,119 @@
+@wasmbus(
+    contractId: "wasmcloud:builtin:sequence",
+    providerReceive: true )
+service Sequence {
```
Sequence is not a service, it's an operation on the `wasmcloud:builtin:extras` contract. I'm okay with changing the contract so that there's a `StartSequence` and `NextSequence` operation, but this is still stuff that belongs as part of an extras contract.

autodidaddict commented on builtin/builtin.smithy:
```diff
@@ -0,0 +1,119 @@
+///
+/// Logging
+///
+@wasmbus(
```
logging and extras, while they can co-exist in the same smithy file if they want, should _not_ be in the same crates. If they require two different signed capabilities, then they must be two different crates because the use of each crate is an implicit demand for permission to use a capability provider that implements that interface.

Today, we declare a dependency on the logging actor interface to get the logging abstraction which requires the actor to be signed with the logging contract ID. The same is true for the extras contract ID. We shouldn't have a single crate that contains the functions for 2 different contracts.

stevelr commented:
> they can co-exist in the same smithy file if they want, should not be in the same crates
- which crate they are in is independent of the contracts. The actor can't talk to a provider unless it's signed with that capability, This crate could split into smaller crates but since they're each so small I thought it made sense to combine them. Keeping them together should reduce compile time for anyone that needs more than one.

- logging, guid, time (to be added later to builtin), have separate contracts so that they can be signed separately, e.g., an actor could have ability to create guids, but not get the time. If they are all in the same builtin:extra crate we don't have that granularity.
-

stevelr commented:
also "Service" in smithy doesn't mean the same thing as a service like we think of http server or rpc server. It's more abstract than that. It's just a collection of operations.

stevelr commented:
... with a version number

autodidaddict commented:
> > they can co-exist in the same smithy file if they want, should not be in the same crates
> 
>     * which crate they are in is independent of the contracts. The actor can't talk to a provider unless it's signed with that capability, This crate could split into smaller crates but since they're each so small I thought it made sense to combine them. Keeping them together should reduce compile time for anyone that needs more than one.
> 
>     * logging, guid, time (to be added later to builtin), have separate contracts so that they can be signed separately, e.g., an actor could have ability to create guids, but not get the time. If they are all in the same builtin:extra contract we don't have that granularity.
> 
> 
> [edited] crate->contract

This is not how it's set up. There's an `extras` contract which includes Guids, Sequences, and Randoms. I'm fine with calling this `generators` instead of extras, but that's the contract. It is a single contract and there will be a single capability provider for it, and there will be a single capability claim in the JWT for it. 

In this PR, you have a contract ID for guid, one for random, and one for sequence number. This needs to be consolidated back into the one "extras" contract.

autodidaddict commented:
re: crates - there should be one actor interface crate for:
* extras (number generators)
* time (when it's created)
* logging

stevelr commented:
- removed logging and sequence
- renamed builtin to numbergen
- both guid and random operations are the same numbergen service, and both use the same contract id

autodidaddict commented on numbergen/rust/src/lib.rs:
```diff
@@ -0,0 +1,4 @@
+//! org.wasmcloud.provider.builtin
```
This module will need full module-level rustdoc if it's going to get published to crates[dot]io. You can use any of the pre-existing actor interfaces as models for this, but at the very least it needs an example of what it looks like to write an actor that uses this interface.

autodidaddict commented:
Also need to fix DCO

autodidaddict commented on numbergen/rust/src/numbergen.rs:
```diff
@@ -0,0 +1,166 @@
+        ctx: &context::Context<'_>,
+        message: &Message<'_>,
+    ) -> Result<Message<'_>, RpcError> {
+        match message.method {
```
These all return scalars and then it looks like we call `serialize` directly on the scalar as the return type. I've tested this with Elixir and you can pack and unpack scalars with no issue... but I'm wondering what happens if Rust is the packer and Elixir is the unpacker. Back in the `widl` days, we had issues where we couldn't round-trip scalars.

@brooksmtownsend do we have any examples of Rust messagepacking Scalars in round trip? In elixir when I messagepack a string scalar, I get `163 | "string"` where 163 is the code. We need to verify that this is what we get out of Rust as well otherwise these scalars won't be compatible with Rust actors.

stevelr replied:
Yes on documentation and unit tests.

Serializing scalars works fine in the testing I've done with rust so far. It's cleaner ergonomics and simpler documentation than using a wrapper, so imo it's worth making it work.

We need a wasmbus interoperability  (w22y) test. Codegen should take care of 80-90% of what we need for other languages. I 'll start with rust plus one or two others.

brooksmtownsend reviewed (changes requested):
one request, and a few comments/questions

autodidaddict commented:
@stevelr this PR still needs the module-level documentation for the crate and the DCO check is still failing.

autodidaddict commented:
@stevelr still waiting for you on this PR.

stevelr commented:
converted to draft pr -need to do some other things before I get back to this

stevelr commented:
these interfaces are completed in the interfaces repo
